docs/developer-docs/security/rust-canister-development-security-best-practices.md

```diff
@@ -581,7 +581,11 @@ Canisters pay for their cycles which makes them inherently vulnerable to attacks
 
 #### Recommendation
 
-Consider monitoring, early authentication, rate limiting on canister level to mitigate this. Also, be aware that an attacker will aim for the call consuming most cycles. See the "Cycle balance drain attacks section" in [how to audit an Internet Computer canister](https://www.joachim-breitner.de/blog/788-How_to_audit_an_Internet_Computer_canister).
+* Consider monitoring, early authentication, rate limiting on canister level to mitigate this. Also, be aware that an attacker will aim for the call consuming most cycles. See the "Cycle balance drain attacks section" in [how to audit an Internet Computer canister](https://www.joachim-breitner.de/blog/788-How_to_audit_an_Internet_Computer_canister).
+
+* For query calls that cause significant computation and don't modify the state, it is advisable to not execute the expensive computation if the method is called as update. Unfortunately, the execution mode of the query (whether it was called as query or update) is currently not directly exposed to the user code. However, one can e.g. call  `ic0.data_certificate_present()` which returns `1` when called as query, and `0` for update methods. See the Interface Specification [section on certified data](/references/ic-interface-spec.md#system-api-certified-data). 
```

Comment:
OTOH this opens up the security issue for the client to get data from a single malicious node. You also (typically?) can't use certified data to mitigate in the case of a computationally heavy query.

Probably still good to mention it here though.

Reply:
Good point, thanks. I mentioned this trade-off and linked to the query certification best practice.